unified_planning/io/pddl_grammar.py

```diff
@@ -0,0 +1,349 @@
+# Copyright 2021-2023 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from functools import partial
+from typing import Dict, List
+import typing
+import unified_planning as up
+import unified_planning.model.walkers
+from unified_planning.io.utils import set_parse_action, set_results_name, Located
+
+import pyparsing
+from pyparsing import ParseResults
+from pyparsing import CharsNotIn, Empty
+from pyparsing import Word, alphanums, alphas, ZeroOrMore, OneOrMore
+from pyparsing import Suppress, Group, Optional, Forward
+
+if pyparsing.__version__ < "3.0.0":
+    from pyparsing import oneOf as one_of
+    from pyparsing import restOfLine as rest_of_line
+else:
+    from pyparsing import one_of
+    from pyparsing import rest_of_line
+
+
+Object = "object"
+TypesMap = Dict[str, unified_planning.model.Type]
+
+
+def nested_expr():
+    """
+    A hand-rolled alternative to pyparsing.nested_expr() that substantially improves its performance in our case.
+    """
+    cnt = Empty() + CharsNotIn("() \n\t\r")
+    nested = Forward()
+    nested <<= Group(
+        Located(
+            Suppress("(") + ZeroOrMore(Group(Located(cnt)) | nested) + Suppress(")")
+        )
+    )
+    return nested
+
+
+NAME = Word(alphas, alphanums + "_" + "-")
+VARIABLE = Suppress("?") + NAME
+PARAMETERS = set_results_name(
+    ZeroOrMore(
+        Group(Located(Group(OneOrMore(VARIABLE)) + Optional(Suppress("-") + NAME)))
+    ),
+    "params",
+)
+
+
+class PDDLGrammar:
+    def __init__(self):
+        # Domain data structures to populate while parsing
+        self.name: str = ""
+        self.requirements: List[ParseResults] = []
+        self.types: List[ParseResults] = []
+        self.constants: List[ParseResults] = []
+        self.predicates: List[ParseResults] = []
```

Comment:
I cannot find the point where you check that this list has length 1 and you throw an error otherwise... You did this for the problem part, but what about the domain? I think only actions can be a "real" list.

Reply:
One alternative to using `Lists` here would be an `Optional` and throwing an error in the patrsing action if the value is not `None` before replacing...

Reply:
I created an `append_unique` method that enforces this behavior directly at parsing time. I tried the `Optional` and `None` approach, creating a unique method that used `getattr` and `setattr` but it was slower than the current approach.